### ISSUE TITLE
Decide on integer datatypes.

### ISSUE THREAD

**Issue by denisalevi** [brian-team/brian2cuda#25]
Our current `CSpikeQueue`([spikequeue.h](https://github.com/brian-team/brian2cuda/blob/72db957/brian2cuda/brianlib/spikequeue.h)) and `cudaVector` ([cudaVector.h](https://github.com/brian-team/brian2cuda/blob/72db957/brian2cuda/brianlib/cudaVector.h)) classes seem to have a random mix of `size32_t`, `int` and `unsigend int` datatypes. Or at least I could not really figure out a consistent reasoning behind it. Now I'm not quite sure what datatypes would make sense here. 

Looking at the `CSpikeQueue` from `brian2` ([cspikequeue.cpp](https://github.com/brian-team/brian2/blob/cdf255f/brian2/synapses/cspikequeue.cpp)), I have a few questions @mstimberg 
- It looks like signed integer types are used for neuron IDs and synapse IDs. **Are we ever handling negative IDs, or is there another reason not to use unsigned integer types?**
- Then for synapse IDs in `CSpikeQueue.synapses`, `size32_t` is used and for synapse IDs in the `CSpikeQueue.queue` `int` is used. **Why not use the same datatype in both?**

My understanding so far is, that the C++ standard requires `unsigned int` to hold only *at least* values in the range `0 to 65535` and `int` in the range `-32767 to 32767`. And the latter seems to me to little for synapse IDs? Or is any computer running `brian2` simulations anyways going to use a computer architecture / compiler where `int` can store a higher range and it simply doesn't really matter?

**Comment by mstimberg** [comment]
We briefly talked about this already, but just to put it in writing: the integer types in Brian 2's C++ code are a bit of a mess, unfortunately. We should certainly unify them at some point.
In general our approach in generated code (so this does not quite apply to the spike queue) is that the model's state variables have types with explicit sizes, e.g. `int32_t` and this type is used everywhere when we generate code for this variable. Explicitly written code in templates should also take this type into account with a construct like this:
```jinja
{%set c_type = c_data_type(variables[varname].dtype) %}
{{c_type}} var = ...
```
*But*, we only did this where we really had to (e.g. a `StateMonitor` can record variables that have double, integer, or boolean types), for synapse indices we simply assume 32bit integers in many places. And then of course we often assume that `int` is 32bit big, which we shouldn't. Finally, the issue between unsigned and signed types is a bit tricky. I think in the beginning I often used the unsigned types for all this variables (indices, etc.) that cannot be negative, but some people with more experience in C++ than me seem to advise against it:
http://wesmckinney.com/blog/avoid-unsigned-integers/
https://google.github.io/styleguide/cppguide.html#Integer_Types

Well, there's definitely room for improvement... From the Brian 2 side, I think it would be good to deal with this at the same time as brian-team/brian2#417.

**Comment by thesamovar** [comment]
Agreed with all this. Do you think it's high priority? Feels like not to me.

**Comment by thesamovar** [comment]
Oops, sorry - thought this was a brian2 issue, please ignore!

**Comment by denisalevi** [comment]
Just referencing [this brian2 merge](https://github.com/brian-team/brian2/pull/814#issuecomment-280158425) here, which tries to avoid `unsigned int` in general. Right now we use `unsigned int` in many places and its still rather messy. So I suppose exchanging unsigned types for signed types everywhere would make sense, following above's advices.

**Comment by denisalevi** [comment]
We are not using `unsigned int` anymore. We still use `int32_t` for some indices and `int` at most other places. But since this seems to not create any problems in brian2 either, we will stick with how it is right now.
